modules/dcache/src/main/java/org/dcache/pool/statistics/IoStatistics.java

```diff
@@ -42,20 +46,25 @@
      *         The duration of the read request in nanoseconds; must be non-negative
      */
 
-    public void updateRead(long readBytes, long readTime) {
+    public void updateRead(long readBytes, long readTime, long requestedBytes) {
 
-        double readSpeed = calculateSpeed(readBytes, readTime);
-        _readSpeeds[_readSpeedsArrayIndex] = readSpeed;
-        _readSpeedsArrayIndex = (_readSpeedsArrayIndex + 1) % _readSpeeds.length;
+        if(readBytes >= 0) {
+            double readSpeed = calculateSpeed(readBytes, readTime);
+            _readSpeeds[_readSpeedsArrayIndex] = readSpeed;
+            _readSpeedsArrayIndex = (_readSpeedsArrayIndex + 1) % _readSpeeds.length;
 
-        _minReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.min(readSpeed, _minReadSpeed);
-        _maxReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.max(readSpeed, _maxReadSpeed);
-        _avgReadSpeed = (_readRequestNum == 0) ? readSpeed : calculateNewAvg(_readRequestNum,_avgReadSpeed, readSpeed);
-        _95ReadSpeed = (_readRequestNum == 0) ? readSpeed : percentile.evaluate(_readSpeeds, 0, (_readSpeedsArrayIndex - 1));
+            _minReadSpeed = (_readRequestNum == 0) ? readSpeed : Math.min(readSpeed, _minReadSpeed);
```

Comment:
One last observation.

Instead of checking the counter each time, you could simply preset/initialize _minReadSpeed to Double.MAX_VALUE and _maxReadSpeed to Double.MIN_VALUE, and then just call Math.min and Math.max.

(I would imagine the apache common code for Percentile would also handle a 0, -1 boundary OK, but that would need checking)

For average, if you examine your method, you should be able to call it without checking (on the first try it would be (0*oldAvg + newSpeed)/1, which equals newSpeed, no matter what oldAvg is, but you could just initialize the average value to 0 as well.

The code does a bit less work this way.